### ISSUE TITLE
Entity Markers

### ISSUE THREAD

**Issue by DasLixou** [bevyengine/bevy#6556]
## What problem does this solve or what need does it fill?

A Marker trait that can mark an entity. example

```rust
#[derive(Marker)]
pub enum Players {
    Red,
    Blue,
    Green,
}
```

then we can assign an entity with a marker
```rust
commands.spawn(PlayerBundle::default()).mark(Players::Red);
```

and we can easily query players
```rust
fn add_items_to_red_and_blue(red_query: Query<Player, Or<MarkedAs<Players::Red>>, MarkedAs<Players::Blue>>>) {
    // ...
}
```

## What solution would you like?

An Marker that works like a Resource but holds a EntityID and the Marker can be used in Queries.

## What alternative(s) have you considered?

Manually make Resources holding EntityIDs and then using a normal query and using .get with the entityID from the resource.

**Comment by harudagondi** [comment]
This requires enum variants as types, which is not possible in Rust. See:

- rust-lang/rfcs#1450
- rust-lang/rfcs#2593
- #4513
- #5639

**Comment by DasLixou** [comment]
> This requires enum variants as types, which is not possible in Rust. See:
> 
>     * [Types for enum variants rust-lang/rfcs#1450](https://github.com/rust-lang/rfcs/pull/1450)
> 
>     * [Enum variant types rust-lang/rfcs#2593](https://github.com/rust-lang/rfcs/pull/2593)
> 
>     * [Indexes: look up entities by the value of their components #4513](https://github.com/bevyengine/bevy/issues/4513)
> 
>     * [Query filter for component value rather than component existence #5639](https://github.com/bevyengine/bevy/issues/5639)

But it works for SystemLabel, and it is infact just a label. the whole entity it corresponds to wouldn't be stored in the marker directly but would be in a lookup table, wouldn't it?

**Comment by harudagondi** [comment]
This specific part would not work: `MarkedAs<Players::Red>`. You cannot use an enum variant as a type.

**Comment by DasLixou** [comment]
> This specific part would not work: `MarkedAs<Players::Red>`. You cannot use an enum variant as a type.

oh! thats pretty sad. But with structs it would work? like
```rust
#[derive(Marker)]
pub struct RedPlayer;
```

?

**Comment by ickk** [comment]
You can basically achieve something to the same effect by just using individual "marker components": 
```rust
#[derive(Component)]
pub struct Red;
#[derive(Component)]
pub struct Blue;
```

```rust
commands.spawn((Player::new(), Red));
```

```rust
fn system(query: Query<Player, With<Red>>) {
    // ...
}
```

**Comment by DasLixou** [comment]
> You can basically achieve something to the same effect by just using individual "marker components":
> 
> ```rust
> #[derive(Component)]
> pub struct Red;
> #[derive(Component)]
> pub struct Blue;
> ```
> 
> ```rust
> commands.spawn((Player::new(), Red));
> ```
> 
> ```rust
> fn system(query: Query<Player, With<Red>>) {
>     // ...
> }
> ```

Yes but that doesn't have to only resolve into one entity. it can have multiple when you don't do it quite well.. but thats probably a better thing than using a resource

**Comment by harudagondi** [comment]
If you need it to resolve it into one entity, why not store that entity in a resource? Like

```rust
#[derive(Resource)]
pub struct RedPlayer(Entity);
#[derive(Resource)]
pub struct BluePlayer(Entity);
```

So that you can guarantee you can have one entity for each color.

**Comment by ickk** [comment]
If you need a singleton but don't want to use a Resource, there is also the [`Query::get_single`](https://docs.rs/bevy/0.8.0/bevy/ecs/system/struct.Query.html#method.get_single) method, which will give you `Result` letting you know if there isn't exactly one match for the query. This won't actually guarantee you only have one, but it will at least let you detect/handle when the invariant is violated.

**Comment by DasLixou** [comment]
yeah but both of them are a little bit messy or don't really guarantee that you'll have one. And both of them make it hard for the scheduler to asynchronously handle them because the information that we only want that one specific entity won't be given to the scheduler (at least thats what i thought)

**Comment by ickk** [comment]
While that's true in some senses.. a system that only mutably queries for a single entity is unlikely to run for a very long period of time. Plus if you had such a large number of systems in your schedule that you were worried about wasting any CPU time at all.. surely many of those other systems would not need access to the same components and so could run concurrently anyway?

**Comment by DasLixou** [comment]
Well.. thats true.. it's just a feature request because i think its a neat feature i would also like. when you say its not worth it then ok :). but ideas are always good, aren't they?

**Comment by ickk** [comment]
certainly :)

**Comment by alice-i-cecile** [comment]
Yep, unless and until we get something like enum-variants-as-types, this can't be implemented. Closing this out :)

**Comment by DasLixou** [comment]
Ok I'm happy to pronounce: I made a working thing for this in 3 hours 🤣. Was very fun to make: https://github.com/ChoppedStudio/bevy_ecs_markers

**Comment by alice-i-cecile** [comment]
Super cool! What's the algorithmic complexity of this?

**Comment by DasLixou** [comment]
> Super cool! What's the algorithmic complexity of this approach?

To get from the marker to the `Entity`, it’s just O(1), but I don’t know how bevy handles Resources and Queries where I pull an `Entity` 😅

**Comment by alice-i-cecile** [comment]
`Query.get` is `O(1)`, and resources are also a constant time lookup :)

**Comment by DasLixou** [comment]
Yesterday I made some performance and style improvements, it now feels like it would be a normal bevy feature 😂

**Comment by DasLixou** [comment]
Because of `EntityNote`s issue, I tried to overhaul the prototype implementation, new release out with bevy 0.13 release
